realm/realm-library/src/main/cpp/jni_util/bcopy.cpp

```diff
@@ -0,0 +1,139 @@
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the University nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+/*
+ * The origin source code has been adapted to fit better into build system
+ * of Realm Java:
+ * - a few type casts have been added
+ * - unused (and undefined) macros have been removed
+ * - added name space realm
+ * - formatted using clang-format
+ */
+
+#include <sys/types.h>
+
+#include "bcopy.hpp"
+
+using namespace std;
+
+namespace realm {
+    namespace jni_util {
+/*
+ * sizeof(word) MUST BE A POWER OF TWO
+ * SO THAT wmask BELOW IS ALL ONES
+ */
+    typedef int word; /* "word" used for optimal copy speed */
+
+#define wsize sizeof(word)
+#define wmask (wsize - 1)
+
+/*
+ * Copy a block of memory, handling overlap.
+ * This is the routine that actually implements
+ * (the portable versions of) bcopy, memcpy, and memmove.
+ */
+
+    void *my_bcopy(void *dst0, const void *src0, size_t length) {
+        char *dst = static_cast<char *>(dst0);
+        const char *src = static_cast<const char *>(src0);
+        size_t t;
+
+        if (length == 0 || dst == src) /* nothing to do */
+            goto done;
+
+/*
+ * Macros: loop-t-times; and loop-t-times, t>0
+ */
+#define TLOOP(s)                                                                                                     \
+    if (t)                                                                                                           \
+    TLOOP1(s)
+#define TLOOP1(s)                                                                                                    \
+    do {                                                                                                             \
+        s;                                                                                                           \
+    } while (--t)
```

Comment:
Are those macro definition correct?

I think those should be in one line like:
```
#define	TLOOP(s) if (t) TLOOP1(s)
#define	TLOOP1(s) do { s; } while (--t)
```